packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -113,51 +108,73 @@ const withInfiniteScroll = (
         showSeeMore: false,
       };
 
-      this.onItemsFetch = props.onItemsFetch;
+      this.props.dataSource.onDataChange(this.updateData);
       this.observer = new IntersectionObserver(this.handleIntersection, {
         threshold: 1,
       });
     }
 
     componentDidMount() {
-      if (this.sentinel) {
-        this.observer.observe(this.sentinel);
-      }
+      this.loadMore().then(newState => {
+        this.setState(newState);
+      });
     }
 
-    componentDidUpdate() {
+    componentDidUpdate(prevProps) {
       if (this.sentinel && this.state.index > 0) {
         this.observer.observe(this.sentinel);
       }
+      if (this.props.dataSource !== prevProps.dataSource) {
+        prevProps.dataSource.removeListener(this.updateData);
+        this.props.dataSource.onDataChange(this.updateData);
+        this.loadMore(0).then(newState => {
+          this.setState(newState);
```

Comment:
We should probably do `setState({ isListFinished: false, ...newState })` here in case the previous dataSource had reached the end of the list.

Reply:
changed